lib/recursive-methods.rb

```diff
@@ -1,49 +1,95 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n) (stack space)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError if !n || (n < 0)
+  
+  return 1 if n <= 1
+  
+  return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError if !s
+  
+  length = s.length
+  return s if length <= 1
+  return s[length - 1] + reverse(s[1...(length - 1)]) + s[0]
```

Comment:
`s[1...(length-1)]` creates a new array and copies all the elements over one-by-one.  So this is actually O(n) on just this line alone.